Python/crossfire/tests/test_crossfire.py

```diff
@@ -10,26 +11,41 @@
     extract_cities_api,
 )
 from crossfire.get_cities import get_cities
-from crossfire.get_fogocruzado import get_fogocruzado
-from decouple import config
+from crossfire.get_fogocruzado import InvalidDateIntervalError, get_fogocruzado
 from geopandas import GeoDataFrame
 from pandas import DataFrame
 
 
-class TestSuccessSignin(TestCase):
-    def setUp(self):
-        fogocruzado_signin(
-            config("FOGO_CRUZADO_EMAIL"), config("FOGO_CRUZADO_PASSWORD")
-        )
+def fake_api_row(**extra_fields):
+    row = {
+        "latitude_ocorrencia": 42.0,
+        "longitude_ocorrencia": 4.2,
+        "densidade_demo_cidade": 42,
+        "uf_estado": "RJ",
+        "nome_cidade": "Rio de Janeiro",
+        "presen_agen_segur_ocorrencia": 1,
+    }
+    if extra_fields:
+        row.update(extra_fields)
+    return row
+
+
+def fake_api_data(*rows):
+    return DataFrame(rows or [fake_api_row()])
 
+
+class TestSuccessSignin(TestCase):
     def test_fogocruzado_signin_environment_variable(self):
         """
         assert the parameters passed to fogocruzado_signin enables FOGO_CRUZADO varibale
         in the environment
         """
+        with patch("crossfire.fogocruzado_signin.get_token_fogocruzado") as mock:
```

Comment:
Essa parte me interessa entender melhor.
Vc esá criando um mock da função `crossfire.fogocruzado_signin.get_token_fogocruzado` e usando ela com qualquer valor, como "FOGOCRUZA_EMAIL" e "FOGOCRUZADO_PASSWORD", e confirmando que essa função tenha sido chamada uma vez.

Reply:
Vamos lá, o teste era assim:

```python
class TestSuccessSignin(TestCase):
    def setUp(self):
        fogocruzado_signin(
            config("FOGO_CRUZADO_EMAIL"), config("FOGO_CRUZADO_PASSWORD")
        )

    def test_fogocruzado_signin_environment_variable(self):
        """
        assert the parameters passed to fogocruzado_signin enables FOGO_CRUZADO varibale
        in the environment
        """
        self.assertTrue(os.environ["FOGO_CRUZADO_EMAIL"])
        self.assertTrue(os.environ["FOGO_CRUZADO_PASSWORD"])
        self.assertTrue(os.environ["FOGO_CRUZADO_API_TOKEN"])
```

O que eu quis mudar (e as razões):
* `setUp` dependia de duas coisas externas, e isso não é bom:
    * Informações sigilosas (senha!) numa variável de ambiente
    * chamda da API real ao executar `fogocruzado_signin`
* No teste em si, que é da `fogocruzado_signin`, esperávamos ter `os.environ["FOGO_CRUZADO_API_TOKEN"]` criado, mas quem cria isso _não_ é a  `fogocruzado_signin` (é `fogocruzado_key`)

Portanto a ideia é fazer o teste garantir que a `fogocruzado_signin` faz o que tem q fazer. Olhando o código dela, ela tem que fazer três coisas (comentei no código dela a seguir):

```python
def fogocruzado_signin(email, password):
    os.environ["FOGO_CRUZADO_EMAIL"] = email. # 1. pegar o valor passado como `email` e salvar como variável de abiente
    os.environ["FOGO_CRUZADO_PASSWORD"] = password. # 2. pegar o valor passado com `password` e salvar como variável de ambiente

    get_token_fogocruzado(). # 3. chamar a função `get_token_fogocruzado`
```

Repare que no ponto 3 eu não preciso _executar_ a função `get_token_fogocruzado`, apenas garantir que ela foi chamada. Por isso um _mock_. Com ele eu consigo verificar se uma função foi chamada.

Quando a gente for _melhorar_ os testes (e não apenas fazer pasar, como descrevi na #8), podemos fazer ele ser ainda melhor verificando os valores:


```python
class TestSuccessSignin(TestCase):
    def test_fogocruzado_signin_environment_variable(self):
        with patch("crossfire.fogocruzado_signin.get_token_fogocruzado") as mock:
            fogocruzado_signin("quarenta", "e dois")
            mock.assert_called_once_with()

        self.assertEqual(os.environ["FOGO_CRUZADO_EMAIL"], "quarenta")
        self.assertEqual(os.environ["FOGO_CRUZADO_PASSWORD"], "e dois")
```